rfc/0015-multi-file-jayvee/0015-multi-file-jayvee.md

```diff
@@ -0,0 +1,160 @@
+
+1. Larger projects will become unmaintainable quite quickly, as the elements cannot be organized into multiple files.
+2. Without distribution to multiple files, there is no possibility to reuse models of other projects.
+
+For example, we might be enable to build libraries of valuetypes that can be reused across multiple projects instead of copying the code.
+
+## Explanation
+
+### Exporting elements for later import
+
+For exporting single elements, I propose to introduce a new keyword `export`.
+All elements within a file are not exportable per default.
+Explicitly declaring an element as exportable allows for later import.
+
+**Example export**
+
+```
+export valuetype MyValueType {
+  // ... details
+}
+```
+
+### Bundling elements to a library for later import
+
+For bundling and exporting elements, I propose introducing a new concept called `libraries`.
+A `library` can inhibit `Valuetype`s, `Block`s, `BlockType`s, `Constraint`s, and `Transform`s.
+A library has to define a version in semver syntax.
+A library has to be exported.
+They serve as entry point to a collection of files (local or remote), similarly to JavaScript's `index.js` mechanism.
+
+**Example library**
+
+```
+export library MyDomainLibrary version 1.2.3 {
+  // definition of a new valuetype as part of the library
+  valuetype MyDomainSpecificValuetype1 {
+    // ... details of valuetype
+  }
+
+  // reference to an existing valuetype to make it part of the library
+  include MyDomainSpecificValuetype2;
+
+  // ... possibly more elements
+}
+```
+
+### Visibility of elements in a file
+
+By introducing the concept of `libraries`, most elements can be defined on three levels in a Jayvee file:
+
+1. On the root level of the file
+2. Within a pipeline
+3. (new) Within a library
+
+The name of an element is given by its definition.
+The **qualified name** is constructed by prepending container structures in this pattern: `<container name>.<element name>`, e.g., `MyDomainLibrary.MyDomainSpecificValuetype1`.
+
+**Access paths:**
+
+- root level elements can access
+  - root level elements by their name
+  - and elements of libraries by their qualified name
+- elements within a library can access
+  - root level elements by their name
+  - and elements of other libraries by their qualified name
+- elements within a pipeline can access
+  - root level elements by their name
+  - elements within the same pipeline by their name
+  - and elements of libraries by their qualified name
+
+**No-access paths:**
+
+- elements within a pipeline cannot be referenced by outside elements
+
+**Imported elements are handled as if they were defined at the root level.**
+
+### Importing elements
+
+Only `export`ed elements can be imported into other files.
+
+#### Importing exported elements of a file
+
+```
+from './path/to/location.jv' use { MyDomainSpecificValuetype1 }; // only imports the defined elements from the file, access via qualified name as if it would be defined at the root level
```

Comment:
Imho `use { Name } from 'path to location'`, then the path can be a long URL and still look nice.

Reply:
Speaking of, this definitely needs clarity on what a valid path is. I'd consider HTTPs urls that lead to jv files (for a lib) valid as well as relative paths to file location (for importing a complete file or lib).

Reply:
+1, similar how deno handles the URL resolving maybe.

```
from './my-local.jv' use { Name }  // Local file
from 'https://jvalue.com/my-org/my-repo' use { Name }  // URL
from 'jv:my-org/my-repo' use { Name }  // Same as above, but one does not need a URL.
```